common/sgx/eeid_verifier.c

```diff
@@ -0,0 +1,301 @@
+        &claims[claims_index++],
+        OE_CLAIM_SIGNER_ID,
+        sizeof(OE_CLAIM_SIGNER_ID),
+        &r_signer_id,
+        sizeof(r_signer_id)));
+
+    OE_CHECK(_add_claim(
+        &claims[claims_index++],
+        OE_CLAIM_PRODUCT_ID,
+        sizeof(OE_CLAIM_PRODUCT_ID),
+        &r_product_id,
+        sizeof(r_product_id)));
+
+    OE_CHECK(_add_claim(
+        &claims[claims_index++],
+        OE_CLAIM_PLUGIN_UUID,
+        sizeof(OE_CLAIM_PLUGIN_UUID),
+        &context->base.format_id,
+        sizeof(oe_uuid_t)));
+
+    *claims_out = claims;
+    *claims_size_out = 7;
+
+    result = OE_OK;
+done:
+    return result;
+}
+
+static oe_result_t _eeid_verify_evidence(
+    oe_verifier_t* context,
+    const uint8_t* evidence_buffer,
+    size_t evidence_buffer_size,
+    const uint8_t* endorsements_buffer,
+    size_t endorsements_buffer_size,
+    const oe_policy_t* policies,
+    size_t policies_size,
+    oe_claim_t** claims,
+    size_t* claims_length)
+{
+    OE_UNUSED(context);
+    OE_UNUSED(policies);
+    OE_UNUSED(policies_size);
+
+    oe_result_t result = OE_UNEXPECTED;
+
+    if (!evidence_buffer || evidence_buffer_size == 0 ||
+        (endorsements_buffer && endorsements_buffer_size == 0))
+        return OE_INVALID_PARAMETER;
+
+    eeid_evidence_t* eeide = (eeid_evidence_t*)evidence_buffer;
+    oe_claim_t* sgx_claims = NULL;
+    size_t sgx_claims_size = 0;
+    uint8_t* sgx_evidence_buffer = eeide->data;
+    size_t sgx_evidence_buffer_size = eeide->eeid_sz;
+    uint8_t* sgx_endorsements_buffer =
+        eeide->endorsements_sz ? eeide->data + eeide->evidence_sz : NULL;
+    size_t sgx_endorsements_buffer_size = eeide->endorsements_sz;
+
+    OE_CHECK(oe_verify_evidence(
+        sgx_evidence_buffer,
+        sgx_evidence_buffer_size,
+        sgx_endorsements_buffer,
+        sgx_endorsements_buffer_size,
+        policies,
+        policies_size,
+        &sgx_claims,
+        &sgx_claims_size));
+
+    const uint8_t* r_enclave_hash = NULL;
+    const uint8_t* r_signer_id = NULL;
+    uint16_t r_product_id = 0;
+    uint32_t r_security_version = 0;
+    uint64_t r_attributes = 0;
+    uint32_t r_id_version = 0;
+
+    for (size_t i = 0; i < sgx_claims_size; i++)
+    {
+        if (strcmp(sgx_claims[i].name, OE_CLAIM_UNIQUE_ID) == 0)
+            r_enclave_hash = sgx_claims[i].value;
+        else if (strcmp(sgx_claims[i].name, OE_CLAIM_SIGNER_ID) == 0)
+            r_signer_id = sgx_claims[i].value;
+        else if (strcmp(sgx_claims[i].name, OE_CLAIM_PRODUCT_ID) == 0)
+            r_product_id = *sgx_claims[i].value;
+        else if (strcmp(sgx_claims[i].name, OE_CLAIM_SECURITY_VERSION) == 0)
+            r_security_version = *sgx_claims[i].value;
+        else if (strcmp(sgx_claims[i].name, OE_CLAIM_ATTRIBUTES) == 0)
+            r_attributes = *sgx_claims[i].value;
+        else if (strcmp(sgx_claims[i].name, OE_CLAIM_ID_VERSION) == 0)
+            r_id_version = *sgx_claims[i].value;
+    }
+
```

Comment:
Should you check if all the values of r_enclave_hash / r_signer_id / r_product_id / r_security_version / r_attributes / r_id_version = 0 are populated? Otherwise, verify_eeid will verify against their init values.

Reply:
In that case `_check_claims` in `attest_plugin.c` would fail and they are also verified further down, so EEID verification would fail if they don't match up. If you like we can add another check here, but I don't think it's necessary.